keps/sig-node/20190530-commit-class.md

```diff
@@ -0,0 +1,405 @@
+multi-tenant or single-tenant clusters can use Pod-oriented APIs, like VPA, to
+achieve good utilization; as such, the `CommitClass` API is likely not
+interesting.
+
+#### Kubernetes deployed on bare metal
+
+As an operator of Kubernetes on bare metal, the compute resources offered by my
+nodes represent a great deal of potential performance. However, picking an
+ideal resource request/limit is very challenging, so individual Pod authors are
+likely to overprovision their Pods. At scale, this results in resource waste
+and poor utilization of the cluster. This problem mirrors that of VM
+infrastructures, where VMs consumers might pick a VM profile with some
+'headroom' relative to workload needs.
+
+`CommitClass` would allow me to reclaim some of that 'headroom' for cluster
+capacity, so that my server purchasing decisions could be driven more by
+utilization than by distributed locally-optimal resource reservation.
+
+#### Kubernetes on VMs, control over hypervisor
+
+As an operator of Kubernetes on VMs with control over the hypervisor settings,
+I have a powerful tool in place to manage server footprint: the hypervisor
+overcommit settings. However, choosing the appropriate hypervisor commit level
+for a Kubernetes cluster is very challenging, because any overcommit level
+applies equally to user workloads and system- or kube- reserved resource
+slices. As a result, attempting to get high density at the VM level can result
+in unstable Kubernetes nodes.
+
+A more Kubernetes-aware commit level would let me isolate overcommit risk to
+user workloads: with `CommitClass` available, I could reduce (or eliminate) VM
+hypervisor overcommit.
+
+#### Kubernetes on VMs, no control over hypervisor / managed Kubernetes (cloud profile)
+
+As an operator of Kubernetes on VMs in the cloud, or of managed Kubernetes,
+I can use Cluster Autoscaling to ensure that my clusters are not provisioned
+wildly out of proportion to Pod resource requests. However, my toolset for
+achieving good utilization of my cluster is limited. A significant gap between
+Pod requests and Pod utilization aross the fleet will result in excessive
+cluster cost, as in other environments.
+
+Quoting from GKE's cluster autoscaling documentation:
+
+> Cluster autoscaler works based on **Pod resource requests**, based on how many
+> resources your Pods have *requested*. Cluster autoscaler does not take into
+> account the resources your Pods are actively *using*. Essentially, cluster
+> autoscaler trusts that the Pod resource requests you've provided are accurate
+> and schedules Pods on nodes based on that assumption.
+
+CommitClass would allow me to manage resource footprint in response to
+utilization, even as Cluster Autoscaler manages resource footprint in response
+to allocation.
+
+### Implementation Details
+
+#### Questions
+
+##### Conflict between multiple CommitClasses
+
+How do you resolve conflict if multiple nodes are selected by the same
```

Comment:
In most areas of our APIs `...Class` types do not select their members, but the individuals reference a class.  E.g. in this case, node would grow a CommitClass field and a defaulting mechanism.

Now you have to deal with whether a class can change in real time (and everyone responds to that in real time) or whether the class needs to be reified into the instance object and the linkage broken.  One is FAR easier to implement and the other is probably more useful.

Reply:
Thanks, the usage of `...Class` as a field individual nodes opt into makes sense.

Didn't totally understand the second point. The quick implementation I put together uses an informer of `CommitClass` objects to dynamically update the resources advertised by kubelet in near-real-time (it reads the informer cache as part of updating the node status, which happens every few seconds). Implementation-wise, this wasn't so bad, but it does have a bunch of consequences to think through.

Jay suggested earlier potentially ditching the `CommitClass` type and extending the Node object spec with resource percentages. If that sounds sane, I think that's probably a stronger approach than introducing a new `...Class`. More of a building block than a framework -- folks who were interested in a higher level API (like the CommitClass proposed here) could write an out-of-tree controller/CRD to manipulate those specs across groups of nodes. Does that seem more reasonable to you?

Reply:
My point was that changing the advertised percentages in real time might (?) require you to back-calculate and change every other pod on the machine.  That might be easy with cgroups, but can you say with confidence that it is true of other implementations?

Reply:
In all APIs, simpler is generally better.  I am not convinced that simply overcommitting everything is the right approach here.  I feel like there are some pods that should not be over-committed, some that should be lightly over-committed, and some that need bold-faced lies.

Reply:
Thank you! I think I follow now. Indeed, I have ideas on how to work real time back-calculation with cgroups. Less certain when it comes to something like kata containers, and Windows is entirely terra incognita.

Totally agree RE segmenting. I've been thinking about that from the perspective of groups of nodes -- e.g., this group of nodes hosts heroes who do excellent capacity planning, and has no over-commit, while this other group of nodes hosts villains who request 20 cores and use 0.2, and is heavily overcommitted.

However, that's fairly indirect, and pushes a good amount of complexity onto the operator (to define these node groups, and then use admission control to segment workloads onto them). A more direct approach would be classifying workloads directly. I'll think about what that might look like.